### PR TITLE
market place api usage implementation

### PR DESCRIPTION
**Pull request by dawba**
Delivers implementation for basic usages of CRUD requests with Controller/Service/Repository architecture.

Suplemented by postman collections for external app usage.

Postman collection json will be updated along the way. Attaching it now, so you can inspect it in the meantime.

[market-place.postman_collection_v2.json](https://github.com/dawba/market-place/files/15314446/market-place.postman_collection_v2.json)

TODO:

- [x] advertisement implementation (rough skeleton without error handling)
- [x] advertisement image implementation
- [x] user implementation
- [x] category implementation
- [x] proper HTTP response codes in returned response bodies
- [x] define proper POST structures for all entities
- [x] proper GlobalExceptionHandler for better error response bodies with messages
- [x] tests for api endpoints

### CODE REVIEW

File: src/main/java/org/marketplace/controllers/AdvertisementManagementController.java [hunk 1]
```diff
@@ -1,22 +1,68 @@
...
+
+    /*
+
+    additional methods that can be implemented in the future:
+
+    - deleting all user ads
+    - deleting all category ads
+    - getting a sum of categories
+
+     */
```
Comment (dimmonn) [inline src/main/java/org/marketplace/controllers/AdvertisementManagementController.java]: just very minor thing
i could suggest to not add //TODO into a production code, might be better to keep as less comments as possible,
it's fine to do that for documentation purpose, like

```
    /**
     * Retrieves a greeting message.
     *
     * @param name the name of the person
     * @return a greeting message
     */
    @GetMapping("/greet")
    public String greet(@RequestParam("name") String name) {
        return "Hello, " + name + "!";
    }

```

    or use a swagger that generates html with doc, like
    
```
import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;
import io.swagger.annotations.ApiParam;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@Api(tags = "GreetingController", description = "Operations pertaining to greetings")
public class GreetingController {

    @GetMapping("/greet")
    @ApiOperation(value = "Get a greeting message", notes = "Provide a name to get a personalized greeting message")
    public String greet(@ApiParam(value = "Name of the person", required = true) @RequestParam String name) {
        return "Hello, " + name + "!";
    }
}
```

it's also totally fine to not use documetation at all.

**Review by dimmonn — COMMENTED** [review]
minor suggestion, might be beneficial to supply new changes with unit tests, otherwise might be hard to not stumble upon regression.

**Comment by dawba** [comment]
I have incorporated suggestions regarding comments with TODOs. 

Added some error handling, however I am stuck when trying to validate multiple fields - it seems like i cannot override basic implementation of ResponseEntityExceptionHandler. Will leave as is and try to deal with it later when rest is ready.

I am aware of the need for supplying the endpoints with the tests - they will be written when I finish the rest of the entity C/R/S system.

**Comment by dawba** [comment]
I have supplied basic tests for Category/Ad/Ad image.

Several tests pass, some of them fail - the reason is that we need User CRS implemented as we require the user to be created to be able to add an advertisement. 

Categories were tested completely - 100% pass rate. With assumption that other methods for Ad image/Ad are pretty much analogous we can expect them to also pass when we finish the user implementation.

@Konrad1w 
Could you follow up on the User implementation as you mentioned couple days before about using encoder and handling all of that logic?

@dimmonn 
Could you tell me if this is how we should be testing the Controllers? Should we also supply the project with tests for Repository/Service?

**Comment by dimmonn** [comment]
> @dimmonn Could you tell me if this is how we should be testing the Controllers? Should we also supply the project with tests for Repository/Service?

sure, it's good practice to test just db layer (you might want to check test containers as an example), but i think having just mockMvc in our project should be fully enough, it indirectly tests the db CRUD operations within db layer 👍
good progress gents 👍😎

File: src/main/java/org/marketplace/services/AdvertisementImageManagementService.java [hunk 2]
```diff
@@ -15,21 +16,34 @@ public AdvertisementImageManagementService(AdvertisementImageManagementRepositor
     }
 
     public AdvertisementImage addImage(AdvertisementImage advertisementImage) {
-        return advertisementImageManagementRepository.save(advertisementImage);
+        try {
+            getAdvertisementImageById(advertisementImage.getId());
+            throw new EntityNotFoundException(String.format("Advertisement image with id: %d already exists!", advertisementImage.getId()));
```
Comment (dimmonn) [inline src/main/java/org/marketplace/services/AdvertisementImageManagementService.java]: hey team, is it some typo? we add image but actually we get image and regardless of the result we fail with an exception..?

Comment (dawba) [reply src/main/java/org/marketplace/services/AdvertisementImageManagementService.java]: Hey, as far as I remember it is deliberate. 

What I wanted to do, is to 

1. Check if the image exists (so to see if created Image instance already exists - if so, throw an error.

Okay at this point I see that it actually is a mistake. The error thrown should be `EntityFoundException`

So yes, I will fix it soon.

The second step should be like:
2. If entity was not wound, then proceed saving the image.
3. If entity was found, propagate the error and finish.

Thanks for catching that @dimmonn

**Comment by Konrad1w** [comment]
Here is completed Postman collection
[Uploading market-place.postman_collection.json…]()

**Pull request merged**